modules/core/src/parallel.cpp

```diff
@@ -775,26 +798,102 @@ static inline int getNumberOfCPUsImpl()
       }
 
    }
-   return cpusAvailable ? cpusAvailable : 1;
+   return cpusAvailable;
+}
+#endif
+
+#if defined CV_LINUX_GROUPS
+static inline unsigned getNumberOfCPUsCFS(void)
+{
+    std::string file_contents = getFileContents("/sys/fs/cgroup/cpu/cpu.cfs_quota_us");
+    if(file_contents.empty())
+        return 0;
+
+    std::istringstream ss_period(file_contents);
+    int cfs_quota;
+    ss_period >> cfs_quota;
+
+    if(cfs_quota < 1) /* cfs_quota must not be 0 or negative */
+        return 0;
+
+    file_contents = getFileContents("/sys/fs/cgroup/cpu/cpu.cfs_period_us");
+    if(file_contents.empty())
+        return 0;
+
+    std::istringstream ss_quota(file_contents);
+    int cfs_period;
+    ss_quota >> cfs_period;
+
+    if(cfs_period < 1)
+        return 0;
+
+    return max(1u, (unsigned)cfs_quota/cfs_period);
 }
 #endif
 
+template <typename T>
+static inline T minNonZero(T val_1, T val_2)
+{
+    return (((val_1 > 0) && (val_2 > 0) && (val_1 > val_2)) || ((val_1 == 0) && (val_2 > 0))) \
+           ? val_2 : val_1;
+}
+
 int cv::getNumberOfCPUs(void)
 {
+    /*
+     * Logic here is to try different methods of getting CPU counts and return
+     * the minimum most value as it has high probablity of being right and safe.
+     * Return 1 if we get 0 or not found on all methods.
+    */
+#if defined CV_CXX11
+    /*
+     * Check for this standard C++11 way, we do not return directly because
+     * running in a docker or K8s environment will mean this is the host
+     * machines config not the containers or pods and as per docs this value
+     * must be "considered only a hint".
+    */
+    unsigned ncpus = std::thread::hardware_concurrency(); /* If the value is not well defined or not computable, returns 0 */
+#else
+    unsigned ncpus = 0; /* 0 means we have to find out some other way */
+#endif
+
 #if defined _WIN32
     SYSTEM_INFO sysinfo;
 #if (defined(_M_ARM) || defined(_M_ARM64) || defined(_M_X64) || defined(WINRT)) && _WIN32_WINNT >= 0x501
     GetNativeSystemInfo( &sysinfo );
 #else
     GetSystemInfo( &sysinfo );
 #endif
+    unsigned ncpus_sysinfo = sysinfo.dwNumberOfProcessors < 0 ? 1 : sysinfo.dwNumberOfProcessors; /* Just a fail safe */
+    ncpus = minNonZero(ncpus, ncpus_sysinfo);
+
+#elif defined CV_LINUX_GROUPS
+    #if defined CV_HAVE_CGROUPS
+        static unsigned ncpus_impl_cpuset = (unsigned)getNumberOfCPUsImpl("/sys/fs/cgroup/cpuset/cpuset.cpus");
+        ncpus = minNonZero(ncpus, ncpus_impl_cpuset);
+
+        static unsigned ncpus_impl_cfs = getNumberOfCPUsCFS();
+        ncpus = minNonZero(ncpus, ncpus_impl_cfs);
+    #endif
+
+    static unsigned ncpus_impl_devices = (unsigned)getNumberOfCPUsImpl("/sys/devices/system/cpu/possible");
```

Comment:
BTW, Why is `/possible` here? (I know that it was here before patch)

See https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-devices-system-cpu:

> online: cpus that are online and being scheduled.
> possible: cpus that have been allocated resources and can be brought online if they are present.

I believe we should use `/online` instead:
```
# cat /sys/devices/system/cpu/online /sys/devices/system/cpu/possible 
0-3
0-3
root ~# echo 0 > /sys/devices/system/cpu/cpu3/online
root ~# cat /sys/devices/system/cpu/online /sys/devices/system/cpu/possible 
0-2
0-3
```

or even both: https://devtalk.nvidia.com/default/topic/1042662/jetson-tx1/-sys-devices-system-cpu-cpu-online-missing-in-the-latest-jetpack-l4t-release/ and select minimal as regular

Reply:
Thanks for the info, I'll read up more on it.

Reply:
This is very true, `/possible` is showing the hardware count, 128 in my case over just `6` for the VM. I saw a few more options and `online` seems to be right. But Android also has CGroups apparently, Can I add it to CGroups list?

Reply:
> Android
> Can I add it to CGroups list?

I believe - yes. It is harmless (just a check for missing files in worst case)